KTIJ-103: Trailing comma breaks formatting of 'when' expression

osipxd opened the pull request:
Issue on YouTrack: [KTIJ-103](https://youtrack.jetbrains.com/issue/KTIJ-103/Trailing-comma-breaks-formatting-of-when-expression)

osipxd commented on plugins/kotlin/formatter/src/org/jetbrains/kotlin/idea/formatter/trailingComma/TrailingCommaState.kt:
```diff
@@ -65,7 +65,7 @@ private fun isMultiline(ktElement: KtElement): Boolean = when {
 
     ktElement is KtWhenEntry -> ktElement.isMultiline(
-        startOffsetGetter = { startOffset },
+        startOffsetGetter = { conditions.lastOrNull()?.endOffset },
```
Enforce trailing comma only in the case there is line break between last condition and arrow.
I'm not sure this fix is correct but iy works.

dimonchik0036 replied:
`startOffset` is correct here. 
Was:
```kotlin
fun t() {
    when (true) {
        <getter scope>true, false,
        true, false </getter scope>-> {}
    }
}
```
After the change:
```kotlin
fun t() {
    when (true) {
        true, false,
        true, false<getter scope> </getter scope>-> {}
    }
}
```

osipxd replied:
Dropped this change

osipxd commented on plugins/kotlin/formatter/src/org/jetbrains/kotlin/idea/formatter/KotlinCommonBlock.kt:
```diff
@@ -1067,7 +1067,16 @@ private val INDENT_RULES = arrayOf(
     strategy("Block in when entry")
         .within(WHEN_ENTRY)
-        .notForType(BLOCK, WHEN_CONDITION_EXPRESSION, WHEN_CONDITION_IN_RANGE, WHEN_CONDITION_IS_PATTERN, ELSE_KEYWORD, ARROW)
+        .notForType(BLOCK, WHEN_CONDITION_EXPRESSION, WHEN_CONDITION_IN_RANGE, WHEN_CONDITION_IS_PATTERN, ELSE_KEYWORD)
```
We need to add indentation to arrow, so I've removed it from this filter.
```diff
 fun foo(x: Any) = when (x) {
     Comparable::class, Iterable::class, String::class, // trailing comma
-    -> println(1)
+        -> println(1)
 
     else -> println(3)
 }
```

osipxd commented on plugins/kotlin/formatter/src/org/jetbrains/kotlin/idea/formatter/KotlinCommonBlock.kt:
```diff
@@ -1067,7 +1067,16 @@ private val INDENT_RULES = arrayOf(
+        .notForType(BLOCK, WHEN_CONDITION_EXPRESSION, WHEN_CONDITION_IN_RANGE, WHEN_CONDITION_IS_PATTERN, ELSE_KEYWORD)
+        .set(Indent.getNormalIndent()),
+
+    strategy("Block after wrapped arrow in when entry")
+        .within(WHEN_ENTRY)
+        .notForType(WHEN_CONDITION_EXPRESSION, WHEN_CONDITION_IN_RANGE, WHEN_CONDITION_IS_PATTERN, ELSE_KEYWORD, ARROW)
+        .forElement { node ->
+            val prevNode = node.prev() ?: return@forElement false
+            prevNode.elementType == ARROW && hasLineBreakBefore(prevNode)
+        }
         .set(Indent.getNormalIndent()),
```
Here I've tried to add indentation after wrapped arrow:
```kotlin
fun foo(x: Any) = when (x) {
    Comparable::class, Iterable::class, String::class, // trailing comma
        -> {
            println(1)
        }
}
```
But I can't get it work. Code always looks like:
```kotlin
fun foo(x: Any) = when (x) {
    Comparable::class, Iterable::class, String::class, // trailing comma
        -> {
        println(1) // <- BLOCK has indent but it is not added to its children 
    }
}
```
Am I trying to do it right way?

osipxd replied:
Fixed it with passing `enforceIndentToChildren = true`. Thanks to @Wamae fo this hint.

dimonchik0036 replied:
I think this change is redundant, because 
```kotlin
fun foo(x: Any) = when (x) {
    Comparable::class, Iterable::class, String::class, // trailing comma
        -> {
        println(1)
    }
}
```
is expected code

osipxd replied:
Removed this strategy

osipxd reviewed (commented):
Also I don't quite sure how it should be formatted in the case when we have line break after arrow. Like this?
```kotlin
when(x) {
    1
        ->
        {
            println(1)
        }
}
```

osipxd commented:
Also I don't quite sure how it should be formatted in the case when we have line break after arrow. Like this?
```kotlin
when(x) {
    1
        ->
        {
            println(1)
        }
}
```

Wamae commented on plugins/kotlin/formatter/src/org/jetbrains/kotlin/idea/formatter/KotlinCommonBlock.kt:
```diff
@@ -1067,7 +1067,16 @@ private val INDENT_RULES = arrayOf(
+            val prevNode = node.prev() ?: return@forElement false
+            prevNode.elementType == ARROW && hasLineBreakBefore(prevNode)
+        }
         .set(Indent.getNormalIndent()),
```
Looking at the code example here in ticket: https://youtrack.jetbrains.com/issue/KTIJ-22247/Trailing-comma-adds-wrong-indentation-in-when-statement
I think the root of the issue is here. the type of indentation used.
What is a `Indent.getNormalIndent()`?
Look at the types on indentations used by the closing brace "}" -> https://github.com/JetBrains/intellij-community/blob/master/platform/code-style-api/src/com/intellij/formatting/Indent.java

Maybe `true` can be passed to `getNormalIndent` so that the indent is relative to the parent.

osipxd replied:
Thank you for the hint! I found the flag `enforceIndentToChildren` and looks like is what i want.

Wamae replied:
Happy to help.

osipxd commented on plugins/kotlin/idea/tests/testData/editor/enterHandler/controlFlowConstructions/ElseInWhenWithoutOption.after.kt:
```diff
@@ -2,8 +2,8 @@ fun some() {
         true && false ->
         else
-        <caret>-> Unit
+            <caret>-> Unit
```
This test-case is the last one thing I can't figure out.
It always fails:
- when I add indentation it says:
   ```
   expected:<...       else
       []<caret>-> Unit
   }...> but was:<...       else
       [    ]<caret>-> Unit
   }...>
   ```
- when I remove indentation it says:
   ```
   expected:<...       else
       [    ]<caret>-> Unit
   }...> but was:<...       else
       []<caret>-> Unit
   }...>
   ```

Wamae replied:
@osipxd Could you share the exact error message?

osipxd replied:
> What does it represent? space? tab or normal_space?
It should be `NORMAL_INDENT`, I think. The problem is that actual result changes when I change expected result (without any changes in code). I'm not quite familiar with these tests so maybe it is expected behaviour. Maybe it runs assertion twice before and after reformatting?

osipxd replied:
@dimonchik0036, can you help, please?

osipxd replied:
The problem persists. This test always fails.

IlyaGulya replied:
This one is fixed in commit 975c2656445ddd234c566a7c70becc35452eade6

dimonchik0036 commented on plugins/kotlin/idea/tests/testData/formatter/trailingComma/whenEntry/WhenEntry.after.kt:
```diff
@@ -13,16 +13,14 @@ fun foo(x: Any) {
         Comparable::class, Iterable::class,
-        String::class, /*// trailing comma*/
-        -> println(1)
+        String::class /*// trailing comma*/ -> println(1)
```
The trailing comma is missing

osipxd replied:
Droped this change

dimonchik0036 commented on plugins/kotlin/idea/tests/testData/formatter/trailingComma/whenEntry/WhenEntry.after.kt:
```diff
@@ -13,16 +13,14 @@ fun foo(x: Any) {
         Comparable::class, Iterable::class,
-        String::class, /*// trailing comma*/
-        -> println(1)
+        String::class /*// trailing comma*/ -> println(1)
```
The same as above, the trailing comma is missing

osipxd replied:
Droped this change

dimonchik0036 commented on plugins/kotlin/idea/tests/testData/formatter/trailingComma/whenEntry/WhenEntry.after.kt:
```diff
@@ -50,17 +48,16 @@ fun foo(x: Any) {
         1, 2,
-        3, /**/
-        -> {
+        3 /**/ -> {
```
The same as above, the trailing comma is missing

osipxd replied:
And this one

dimonchik0036 commented on plugins/kotlin/idea/tests/testData/formatter/trailingComma/whenEntry/WhenEntry.after.kt:
```diff
@@ -69,8 +66,7 @@ fun foo(x: Any) {
         1, 2,
-        3, /**/
-        -> {
+        3 /**/ -> {
```
The same as above, the trailing comma is missing

osipxd replied:
Droped this change

dimonchik0036 commented on plugins/kotlin/idea/tests/testData/formatter/trailingComma/whenEntry/WhenEntry.after.inv.kt:
```diff
@@ -20,8 +20,7 @@ fun foo(x: Any) {
         Comparable::class, Iterable::class,
-        String::class, /*// trailing comma*/
-        -> println(1)
+        String::class /*// trailing comma*/ -> println(1)
```
The trailing comma is missing

IlyaGulya commented:
@osipxd Hi!
Are you going to continue work on this PR?
I'd like to fix this issue

osipxd commented:
@IlyaGulya, feel free to work on it! I've rebased the branch onto the fresh master and fixed review comments. If you want continue work in this branch, let me know and I'll give you acces to the fork.

IlyaGulya commented:
@dimonchik0036 
Hi! I've just pushed fixes for this issue.
Also, I've added separate test `WhenEntryArrow` which checks that formatter correctly behaves on caret moving to next line by Enter action.
Please take a look and tell if everything is ok.

IlyaGulya commented:
@dimonchik0036 Hi!
Please, take a look on this PR, or, at least, ping someone who is responsible for it now.

IlyaGulya commented:
@vladimirdolzhenko Hi!
Can you please also take a look on this PR? 🙂

IlyaGulya commented:
@dimonchik0036 @vladimirdolzhenko
Gentle reminder about this PR 🙂 
I've rebased on latest master

IlyaGulya commented:
@aleksandrina-streltsova thanks!
